sessionmanager.md

```diff
@@ -0,0 +1,161 @@
+  body: {
+    mode: "edit",
+    content: "content of the edit"
+  }
+}
+```
+
+Multiple edits or transformations could be sent for a document, at any time.
+
+To handle this the user would need to write code similar to this:
+
+```javascript
+while (true) {
+  // some sort of check to make sure we're not processing
+  // too many sessions at a time. The odds of having more documents
+  // than available processors is _high_.
+  acquireSemaphore();
+
+  // by not specifying a session id we will get the next unlocked session from service bus.
+  const sessionReceiver: SessionReceiver = serviceBusClient.getSessionReceiver(
+    "queue",
+    "peekLock"
+  );
+  startSessionLockAutoRenewal(sessionReceiver);
+
+  sessionReceiver.subscribe(messageHandlers);
+
+  // ...after some predefined period of time
+  sessionReceiver.close();
+
+  stopSessionLockRenewal(sessionReceiver);
+  releaseSemaphore();
+}
+```
+
+## Proposal
+
+In the same vein as what we provide in `Receiver.subscribe()` we can provide a
+`SessionManager` that would support the use case where the user would like
+to read messages, in parallel, from a number of sessions while controlling
+the amount of parallelism.
+
+This would simplify session usage to the point where it becomes more accessible,
+allowing more users to use this powerful feature.
+
+The proposed interface for this change:
+
+```javascript
+class ServiceBusClient {
+  // a session manager is conceptually a receiver, but does not have to
+  // obey the same interface.
+  createSessionManager("queue", "peekLock", sessionManagerOptions): SessionManager;
+  createSessionManager("topic", "subscription", "peekLock", sessionManagerOptions): SessionManager;
+}
+
+interface SessionManagerOptions {
+  // amount of time we will wait for messages to arrive in a session before
+  // it is `close()`d
+  idleTimeoutMs: number;
+
+  // the maximum number of receivers that can be active at any
+  // time. Allows the user to avoid getting steamrolled by too many
+  // sessions.
+  maxConcurrentReceivers: number;
+
+  // new feature (and probably pushed down to SessionReceiver as well) to enable auto-renewal
+  // of the session lock.
+  maxSessionAutoRenewLockDurationInMs?: number;
+
+  // quite a few of the options available in MessageHandler (used for normal Receiver subscribe)
+  // would also apply here.
+  autoComplete?: boolean;
+}
+
+// implicit to this - this is specific to a single session.
+interface SessionContext {
+  // managing session state is implicit to the context where it is called. This flows
+  // similarly to what developers would do with `SessionReceiver`
+  setState();
+  getState();
+
+  sessionId: string;
+}
+
+interface SessionMessageHandler {
+  // tentatively suggesting proposal #2 for error handling below.
+  processMessage(message: ReceivedMessage|ReceivedMessageWithLock, context: SessionContext);
+  processError(err: Error, context: { sessionId: string });
+}
+
+interface SessionManager {
```

Comment:
I would like to be able to get the number of active sessions the SessionManager is managing, as well as callbacks for when a session is reserved or released.

It would be helpful for telemetry and scaling purposes. I will want to monitor how well the session managers are keeping up with the amount of sessions in the system so I can detect/alert when more session managers need to be created.

Reply:
@JoshLove-msft , @hemanttanwar , @KieranBrantnerMagee  - any objections to this idea on y our end?

CC: @ramya-rao-a

Reply:
`get the number of active sessions` : That is fine. 
But a callback when session is reserved, I am not right away excited about it because the purpose of SDK making it easy without user worrying about it. There are other ways to get telemetry and system performance. Now you will have three call backs 1. message received, 2. error call back 3. session reserved call back.

Reply:
@hemanttanwar  - funnily after talking with @ramya-rao-a  we're thinking that we'd **definitely** do the callbacks for reserve/release and not export the active number (primarily because that can easily be derived using the other callbacks).

The session reservation and release exist in a similar form in EventHubs (as partitionInitialize/partitionClose) and the goal there is to say "go ahead and init your db connection or whatever resource" and then give them both a natural spot to both open and close it. So I think they make sense from that perspective. I think of it as making things easier for the user, otherwise they would need some way of tracking that same concept _somehow_. 

Given that we funnel all messages into the same callback that wouldn't necessarily be trivial.

Reply:
For .NET, we will most likely be adding one for session released/complete - https://github.com/Azure/azure-sdk-for-net/issues/11349

I think it makes sense to have one for session accepted as well. We didn't have anyone specifically asking for this and figured it could be added in a non-breaking way in the future. Based on your and @GaryWilber comments, we can consider doing the reserve/accept event as well.